frontend/src/metabase/query_builder/components/NativeQueryEditor.jsx

```diff
@@ -269,11 +270,21 @@ export default class NativeQueryEditor extends Component {
       showLineNumbers: true,
     });
 
+    this._lastAutoComplete = { timestamp: 0, prefix: null, results: null };
+
     aceLanguageTools.addCompleter({
       getCompletions: async (editor, session, pos, prefix, callback) => {
         try {
-          // HACK: call this.props.autocompleteResultsFn rather than caching the prop since it might change
-          const results = await this.props.autocompleteResultsFn(prefix);
+          let { results, timestamp } = this._lastAutoComplete;
+          const cacheHit =
+            Date.now() - timestamp < AUTOCOMPLETE_CACHE_DURATION &&
+            this._lastAutoComplete.prefix === prefix;
+          if (!cacheHit) {
+            // HACK: call this.props.autocompleteResultsFn rather than caching the prop since it might change
+            results = await this.props.autocompleteResultsFn(prefix);
+            this._lastAutoComplete = { timestamp: Date.now(), prefix, results };
+          }
```

Comment:
What if we abstract this away, so we don't get lost in the implementation too much. I wish I could just read something like

```ts
const results = await this._autocompleteResultsFnWithCache(prefix)
```

or something similar

Reply:
I thought about that, but then since this is just a hack until we got a new Ace-based component in place, likely with a rewritten autocomplete logic anyway, then it is also beneficial to let this hack appears front and center.